### DISCUSSION TITLE
Extend type programatically

### DISCUSSION THREAD

**Discussion by dennis-koster** [nuwave/lighthouse#2617] [Q&A]
Hi,

I am creating a custom directive that can be applied on a type definition. The aim is for it to generate an additional type, based on the type the directive is applied to and then extend the original type with an attribute that refers to the generated type.

I have managed to generate the new type programatically using the type registry in the `manipulateTypeDefinition` method to register a new `ObjectType`. However, I am am struggling to find a way to register a type extension.

Here is what I am trying to achieve in a nutshell. Given this schema:

```graphql
type NewsItem @translatable {
    id: ID!

    title: TranslatableString!
    intro: TranslatableString
    content: TranslatableString
}
```
I want the following to be generated:
```graphql
# This I have managed
type NewsItemTranslation {
    locale: String!
    title: String!
    intro: String
    content: String
}

# This I need help with
extend type NewsItem {
    translations: [NewsItemTranslations!]!
}
```

Any nudge in the right direction would be hugely appreciated!

**Comment by dennis-koster** [comment]
UPDATE: I have now, somewhat, managed to add the `translations` field to the type defintion of `NewsItem` with the code below. However, I can not image this being the correct way with all of these nested node instantiations.

```php
public function manipulateTypeDefinition(
    DocumentAST &$documentAST,
    TypeDefinitionNode &$typeDefinition
): void {
    $typeName = $typeDefinition->getName()->value;
    /** @var ObjectTypeDefinitionNode $rootType */
    $rootType = $documentAST->types[$typeName];

    $rootType->fields = ASTHelper::prepend(
        $rootType->fields,
        new FieldDefinitionNode([
            'name'       => new NameNode([
                'value' => 'translations',
            ]),
            'type'       => new NonNullTypeNode([
                'type' => new ListTypeNode([
                    'type' => new NonNullTypeNode([
                        'type' => new NamedTypeNode([
                            'name' => new NameNode([
                                'value' => $typeName . 'Translation',
                            ]),

                        ]),
                    ]),
                ]),
            ]),
            'directives' => new NodeList([]),
            'arguments'  => new NodeList([]),
        ]),
    );
}
```

**Comment by stayallive** [comment]
You can take some inspiration from the `PaginationManipulator` class since it does a similar transformation as you are trying to achieve: 

https://github.com/nuwave/lighthouse/blob/0fe0f924a0e72a87ac726296bb95f41f2029a2ab/src/Pagination/PaginationManipulator.php#L96-L105

Especially using `Parser::objectTypeDefinition` (and friends) might be interesting to create those objects from GraphQL code.

**Reply by dennis-koster** [reply]
Great find! I'll see if I can get to work, thanks for the tip!
